From 6c351faf55a3eae75cba059bf40c43892824311d Mon Sep 17 00:00:00 2001
From: egg82 <eggys82@gmail.com>
Date: Thu, 8 Aug 2019 21:55:41 -0600
Subject: [PATCH] Add village API

This adds a Bukkit (Paper) API to get and modify info on villages and sieges in the world.
There's a fair number of changes here because this update completely changes how villages work.

diff --git a/src/main/java/com/destroystokyo/paper/world/Siege.java b/src/main/java/com/destroystokyo/paper/world/Siege.java
new file mode 100644
index 00000000..3a4b9962
--- /dev/null
+++ b/src/main/java/com/destroystokyo/paper/world/Siege.java
@@ -0,0 +1,13 @@
+package com.destroystokyo.paper.world;
+
+import org.bukkit.Location;
+
+public interface Siege {
+    SiegeState getState();
+
+    int getWavesRemaining();
+
+    int getMobsRemaining();
+
+    Location getMobSpawnLocation();
+}
diff --git a/src/main/java/com/destroystokyo/paper/world/SiegeState.java b/src/main/java/com/destroystokyo/paper/world/SiegeState.java
new file mode 100644
index 00000000..7f9434a2
--- /dev/null
+++ b/src/main/java/com/destroystokyo/paper/world/SiegeState.java
@@ -0,0 +1,8 @@
+package com.destroystokyo.paper.world;
+
+public enum SiegeState {
+    TONIGHT,
+    CAN_ACTIVATE,
+    ACTIVATED,
+    DONE
+}
diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index 1e7b633c..aa103928 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -1,5 +1,6 @@
 package org.bukkit;
 
+import com.destroystokyo.paper.world.Siege;
 import java.io.File;
 import org.bukkit.generator.ChunkGenerator;
 
@@ -2159,6 +2160,19 @@ public interface World extends PluginMessageRecipient, Metadatable {
      */
     public double getHumidity(int x, int z);
 
+    // Paper start - village API
+
+    /**
+     * Tries to activate a random siege in the world.
+     * This requires a player to be nearby a village.
+     *
+     *
+     * @return
+     */
+    @Nullable
+    Siege tryActivateRandomSiege();
+    // Paper end
+
     /**
      * Gets the maximum height of this world.
      * <p>
-- 
2.22.0.windows.1

